packages/@cdktf/provider-generator/lib/get/generator/provider-schema.ts

```diff
@@ -10,14 +10,35 @@ import { ConstructsMakerProviderTarget } from "../constructs-maker";
 
 const terraformBinaryName = process.env.TERRAFORM_BINARY_NAME || "terraform";
 
+/**
+ * Fully Qualified provider name in the format:
+ * like e.g. registry.terraform.io/hashicorp/aws
+ */
+export type FQPN = string & { __type: "FullyQualifiedProviderName" };
+export type ProviderHostname = string & { __type: "ProviderHostname" };
+export type ProviderNamespace = string & { __type: "ProviderNamespace" };
+export type ProviderName = string & { __type: "ProviderName" };
+
+export const parseFQPN = (f: FQPN) => {
+  const [hostname, namespace, name] = f.split("/");
+  if (!name) {
+    throw new Error(`can't handle ${f}`);
+  }
+  return { hostname, namespace, name } as {
+    hostname: ProviderHostname;
+    namespace: ProviderNamespace;
+    name: ProviderName;
+  };
+};
+
 export interface ProviderSchema {
   /*
   0.1 is e.g. returned by Terraform 0.14
   0.2 is e.g. returned by Terraform 1.0 (0.2 added support for nested_type / plugin protocol v6)
   */
   format_version?: "0.1" | "0.2";
-  provider_schemas?: { [type: string]: Provider };
-  provider_versions?: { [fqn: string]: string };
+  provider_schemas?: { [fqpn: string]: Provider };
```

Comment:
wouldn't this be 

```suggestion
  provider_schemas?: { [fqpn: FQPN]: Provider };
```

then?

Reply:
Yeah! I tried that but got errors when trying to build 😞 Even though there were no errors in VS Code 🧐

Reply:
```
provider-generator git:(fix-provider-docs-links) ✗ yarn build                                                                      17:55:02
yarn run v1.22.17
$ tsc
lib/get/generator/provider-generator.ts:88:29 - error TS7053: Element implicitly has an 'any' type because expression of type 'FQPN' can't be used to index type '{}'.

88     const providerVersion = this.schema.provider_versions?.[fqpn];
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

lib/get/generator/provider-generator.ts:109:22 - error TS7053: Element implicitly has an 'any' type because expression of type 'FQPN' can't be used to index type '{}'.

109     const provider = this.schema.provider_schemas?.[fqpn];
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

lib/get/generator/provider-generator.ts:116:47 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Schema'.

116         this.resourceParser.parse(fqpn, type, resource, "resource")
                                                  ~~~~~~~~

lib/get/generator/provider-generator.ts:121:57 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Schema'.

121         this.resourceParser.parse(fqpn, `data_${type}`, resource, "data_source")
                                                            ~~~~~~~~

lib/get/generator/provider-generator.ts:137:22 - error TS7053: Element implicitly has an 'any' type because expression of type 'FQPN' can't be used to index type '{}'.

137     const provider = this.schema.provider_schemas?.[fqpn];
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

lib/get/generator/provider-schema.ts:40:25 - error TS1023: An index signature parameter type must be either 'string' or 'number'.

40   provider_schemas?: { [fqpn: FQPN]: Provider };
                           ~~~~

lib/get/generator/provider-schema.ts:41:26 - error TS1023: An index signature parameter type must be either 'string' or 'number'.

41   provider_versions?: { [fqpn: FQPN]: string };
                            ~~~~


Found 7 errors.
```

Reply:
(the last two ones are directly related)

Reply:
Requires TS 4.4: https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#symbol-template-signatures